### ISSUE TITLE
Mortalとlisjong Policyの同一局面decision disagreement診断を追加する

### ISSUE THREAD

**Issue by lisbun** [lisbun/lisjong-arena#118]
## Background

`lisjong` のrule-based強化について、次の改善対象をheuristicで決め打ちするのではなく、強いreference policyであるMortalとの**同一局面decision disagreement**を観測し、今後のrule-based / ML強化テーマを実証的に選びたい。

現在Arenaには以下が成立している。

- `#67`
  - `MortalMixedGameRunner`
  - Mortal 1席 + lisjong Policy 3席
  - fixed seed / `4p-red-single` / Mortal seat rotation
- `#116` / PR `#117`
  - Mortal evaluationのbaselineをregistered Policyへ一般化
- `#55`
  - standard `LocalGameRunner` にはopt-in decision observation compositionがある
  - ただしMortal mixed executionは明示的にscope外

current main at issue creation:

```text
2672cc24b90712e98d863727e3bb55785035c35b
```

現行 `MortalMixedGameRunner` ではMortal seatについて、

```text
Observation.new_events()
    -> Mortal
    -> MJAI response
    -> Observation.select_action_from_mjai()
    -> RiichiEnv Action
```

を実行し、そのactionだけをgameへ適用している。

一方、同じMortal-seat `Observation` に対してlisjong Policyをshadow実行すれば、game trajectoryを分岐させずに、

```text
same seat-relative observation
    ├─ Mortal        -> action M  # applied
    └─ lisjong Policy -> action L  # shadow only
```

というpaired decision comparisonを取得できる。

本Issueではこの最小診断pathを追加する。

---

## Goal

Mortal mixed single-round executionに、**Mortal seatだけを対象とするopt-in / dedicated shadow comparison path** を追加し、同一decision stateに対する

- Mortal selected action
- selected lisjong Policy action
- agreement / disagreement
- lisjong側のplayer-safe decision context / trace

を対応付けてinspectできるようにする。

重要な原則は、

> **Mortal actionだけがRiichiEnvへ適用され、shadow lisjong actionはgame progressionへ一切影響しない**

ことである。

Mortalをground truthとは扱わない。

---

## Why shadow comparison on the Mortal seat

単純にMortalとlisjongを別gameで走らせ、action列を比較する方法はprimary diagnosticにしない。

最初にactionが異なった時点でgame stateが分岐するため、その後の差が

- 同一局面でのdecision差
- 過去action差によるstate差

のどちらなのか判別できなくなるためである。

本Issueではexisting Mortal mixed executionをdriverとして維持し、Mortalがactionを要求されたexact same `Observation` にlisjong Policyをshadow実行する。

既存ABBB/Mortal seat rotationを利用すれば、各rotationでtarget seatをSeat 0..3へ移動できる。

---

### 1. Existing game semanticsを変えない

shadow comparison on/offで、Mortal mixed gameのobjective executionを変えない。

少なくとも以下を維持する。

```text
Mortal action is the only applied action on Mortal seat
baseline Policy 3 seats use existing execution path
fixed seed
4p-red-single
Mortal seat rotation
RoundStats
final scores / ranks
fail-closed Mortal runtime semantics
serial Mortal execution
```

shadow lisjong Policyのselected actionを`env.step()`へ渡してはならない。

shadow比較のためにMortalを二重実行しない。

---

### 2. Same observation / same legal-action boundary

Mortalとshadow Policyは、Mortal seatに対する同じRiichiEnv `Observation` から判断する。

shadow Policy側はexisting Arena-owned Adapter seamを再利用する。

概念的に:

```text
same Observation
    |
    |-- Mortal path
    |     new_events()
    |       -> Mortal runtime
    |       -> MJAI response
    |       -> select_action_from_mjai()
    |       -> applied RiichiEnv Action
    |
    `-- shadow lisjong path
          build_decision()
            -> PolicyInput / mapping
          execute_policy_with_trace()
            -> canonical InternalAction / DecisionTrace
          mapping.resolve()
            -> shadow RiichiEnv Action
```

Arenaがlisjong `PolicyInput` / `InternalAction` / `DecisionTrace` semanticsを再定義しない。

hidden state / oracle informationをshadow Policyへ渡さない。

---

### 3. Dedicated shadow Policy runtime

Mortal seat用に、比較対象lisjong Policyの独立runtime stateを持つ。

少なくともexisting Adapterと同じ、

```text
Policy
SeatMaterializedState
RiichiEnvActionMappingSession
```

を利用する。

他3席のactual baseline Policy instance / tracker / mapping sessionをshadow比較用に再利用しない。

Policy instance lifecycleやmutable stateをseat間で共有しない。

---

### 4. Comparison semantics

Python object identity、`repr()`、単純なprotocol文字列だけをagreement判定に使用しない。

両者が同一Observation上でlegal RiichiEnv Actionへ解決された後、**semantic external action**としてnormalization / comparisonする。

少なくとも、meaningfully identicalなactionはagreementとなり、異なるactionはdisagreementとなること。

normalizerはpurpose-specificかつ最小にし、generic cross-environment Action schemaを新設しない。

Mortal responseがmalformed / illegal、またはshadow Policy actionがlegal external Actionへresolveできない場合は、通常disagreementとして数えずfailureとしてfail closedまたは明示的failure outcomeにする。

既存Mortal fail-closed semanticsを弱めない。

---

### 5. Structured decision comparison record

exact class名はcurrent namingに合わせてよいが、各Mortal-seat decisionについてimmutableなstructured recordを保持する。

最低限:

```text
seed
rotation / target seat
step or monotonic decision identity
reference Policy identity

lisjong PolicyInput
lisjong DecisionTrace

normalized Mortal selected action
normalized lisjong selected action

agreement: bool
```

必要ならraw Mortal responseをdiagnostic fieldとして保持してよいが、credential / container configuration等を混ぜない。

Mortalのinternal value / logits / hidden representationを取得することはrequirementにしない。

`PolicyInput` / `DecisionTrace`はexisting lisjong-owned typed valueをtransportし、Arena側でdict semanticを再発明しない。

---

### 6. Applied-vs-shadow identityを明示する

record / result APIから、

```text
driver = Mortal
shadow = selected lisjong Policy
```

であることが曖昧にならないこと。

`agreement == false` を「Mortalが正しい」「lisjongが誤り」と表現しない。

---

### 7. Diagnostic aggregation

1 runまたは複数rotationのresultから最低限以下を取得できるようにする。

```text
total paired decisions
agreements
disagreements
agreement rate
```

さらに、既存typed/public informationから安全かつ低コストで導出できる範囲で、少なくともaction-kind pairのbreakdownを提供する。

例:

```text
Mortal discard / lisjong discard
Mortal discard / lisjong riichi-related action
Mortal call / lisjong pass
...
```

exact categoriesはcurrent RiichiEnv action semanticsに合わせる。

打点・押し引き・牌効率等の**原因ラベルを自動的に付けない**。

---

### 8. Representative disagreement inspection

全decisionを人間が読む前提にしない。

disagreementだけを後から抽出できるAPIを提供する。

最低限、

```text
all disagreements
first N disagreements
filter by normalized action kind where practical
```

程度のinspectionを可能にする。

初期実装でgeneric viewer / dashboardを作らない。

永続化については、existing safe/versioned serializerを自然に再利用できる場合のみpurpose-specificなopt-in diagnostic artifactを追加してよい。

**このIssueのためだけに `PolicyInput` / `DecisionTrace` のgeneric persistence schema、canonical GameRecord、databaseを新設しない。**

machine-readable persistenceがgeneric schema拡張を要求する場合は、まずin-memory structured result + reproducible CLI summary / extractionまでを完成させ、必要性をfollow-upとして報告する。

---

## CLI / execution entry point

existing strength comparison outputへ大量のdecision diagnosticを常時混ぜない。

以下のどちらかを優先する。

1. dedicated diagnostic CLI
2. existing Mortal CLIの明示opt-in diagnostic mode

conceptual example:

```powershell
python -m lisjong_arena.mortal_decision_compare `
  --policy combined `
  --seeds 0:24 `
  --mortal-image "mortal@sha256:<digest>" `
  --mortal-revision "<revision>" `
  --mortal-model "C:\path\to\model.pth"
```

exact module / option名はcurrent CLI architectureへ合わせてよい。

Mortal runtime/model provenance requirementはexisting Mortal evaluation contractを再利用する。

---

## Initial diagnostic run

implementation完了後にsmall real-Mortal diagnosticを実行する。

まずはbounded runで、

```text
fixed seeds
Mortal seat rotation
selected current baseline/reference Policy
```

についてpaired recordsが実際に取得できることを確認する。

その後、次の強化テーマを選ぶための探索runとして**約10,000 paired decisionsを目安**に利用できる設計を目標とするが、10,000件の実行自体をmerge acceptance conditionにはしない。

Mortal inference cost / runtimeを実測してからanalysis populationを決める。

---

## Tests

実Mortal modelを通常unit testへ要求しない。

focused fake runtime / fake Observation等を優先する。

最低限:

1. same Mortal-seat ObservationがMortal path / shadow Policy pathの双方へ使われる
2. shadow Policyはexactly once実行される
3. Mortal actionだけが`env.step()`へ適用される
4. shadow actionがgame progressionへ流入しない
5. shadow Policy用runtimeがactual 3 baseline seatsと独立している
6. shadow Policyの`PolicyInput.self_seat` / selected action actorがMortal seatと一致する
7. semantically identical external actions -> agreement
8. semantically different actions -> disagreement
9. representation差だけのsame actionをfalse disagreementにしない
10. malformed / illegal Mortal responseをordinary disagreementへ丸めない
11. shadow Policy mapping failureをordinary disagreementへ丸めない
12. comparison on/offでsame Mortal responsesを与えたobjective game resultが同一
13. agreement/disagreement aggregationがdeterministic
14. disagreement extractionが元recordを失わない
15. existing Mortal evaluation testsがgreen
16. existing Policy-vs-Policy / standard LocalGameRunner behaviorがgreen

実Mortal smokeでは少なくとも1 fixed seed / 4 rotationsについて、paired decision count > 0 とsummary生成を確認する。

---

## Documentation

Arenaのrelevant evaluation documentationへ、Mortal decision comparisonを

> **strong referenceとのsame-state diagnostic**

として記載する。

明記すること:

- Mortalはground truthではない
- disagreementはerrorではない
- Mortal actionだけがdriver
- lisjong actionはshadow
- purposeは次のrule-based / ML課題を選ぶためのdiagnosis
- strength benchmarkとは別のmeasurementである

---

## Non-goals

本Issueでは以下を行わない。

```text
Mortal imitation learning
Mortal actionを教師ラベルにしたtraining dataset作成
learned Policy実装
HandBelief consumer変更
rule-based Policy強化

disagreement原因の自動分類
「打点」「押し引き」等のhuman-semantic label自動付与
Mortal action superiorityの証明
counterfactual rollout

Mortalを4席同時shadow実行
Mortal複数processで全Policy seatを同時比較
Mortal parallelization

canonical decision corpus
project-wide GameRecord
PolicyInput / DecisionTrace schema expansion
generic persistence framework
SQLite / DB / dashboard / viewer

generic external-agent API
generic backend abstraction
```

---

## Stop / re-scope triggers

以下が必要と判明した場合は、本Issue内でarchitectureを膨らませず報告する。

- same Mortal-seat Observationをshadow Policyへ安全にmaterializeできない
- comparisonのためにlisjong Policy contract変更が必要
- semantic external action比較のためにgeneric Action schema新設が必要
- representative case inspectionのためにcanonical historical persistence設計が必要
- one Mortal runtime/gameというexisting lifecycleを壊す必要がある
- hidden informationを使わないと比較contextを構築できない

---

## Acceptance Criteria

- [ ] current main / `AGENTS.md` / #55 / #67 / #116/#117をpreflightで再確認している
- [ ] Mortal mixed executionにopt-in/dedicated same-state shadow comparison pathがある
- [ ] same Mortal-seat `Observation` からMortalとselected lisjong Policyの双方を評価する
- [ ] Mortal actionだけがRiichiEnvへ適用される
- [ ] shadow Policy actionはgame stateへ影響しない
- [ ] Mortal seat用の独立shadow Policy runtimeを利用する
- [ ] existing Adapter / Policy execution semanticsを再利用する
- [ ] hidden state / oracle informationをshadow Policyへ渡さない
- [ ] semantic external actionとしてagreement / disagreementを判定する
- [ ] representation差だけでfalse disagreementを作らない
- [ ] structured immutable comparison recordを取得できる
- [ ] lisjong `PolicyInput` / `DecisionTrace`をtyped valueのまま保持する
- [ ] total / agreement / disagreement / agreement rateを集計できる
- [ ] basic action-kind breakdownを確認できる
- [ ] disagreementだけを抽出・inspectできる
- [ ] Mortalをground truthとして扱わない
- [ ] comparison on/offでobjective Mortal mixed-game semanticsが変わらない
- [ ] standard `LocalGameRunner` / normal Policy comparisonを不要に変更していない
- [ ] generic persistence / canonical GameRecord / DB / dashboardを導入していない
- [ ] real Mortal fixed-seed smokeでpaired decision recordとsummaryを確認している
- [ ] focused tests / full CI / Ruff / `git diff --check` がgreen

---

## Follow-up use

本Issue完了後、このdiagnosticを使ってMortalとlisjongのdisagreement分布を観測する。

次の強化方向は事前に固定しない。

候補例:

```text
offensive value / 打点
riichi / dama
call / pass
defense / push-fold
placement / score context
その他
```

観測結果から上位のcapability gap仮説を立て、rule-basedで直すべきか、HandBelief / value model / learned Policy等のML側へ進むべきかを別途判断する。

## Related

- #55 — standard LocalGameRunner decision observation composition
- #67 — Mortal mixed-agent single-round evaluation
- #116 / PR #117 — Mortal baseline generalization
